### ISSUE TITLE
Error handling long Postgres table names

### ISSUE THREAD

**Issue by jamey-clari** [liquibase/liquibase#1483]
## Environment
**Liquibase Version**: 4.0.0

**Liquibase Integration & Version**: Java API

**Liquibase Extension(s) & Version**: 

**Database Vendor & Version**: Postgres 10.12

**Operating System Type & Version**: Mac + Linux

## Description
If I create a simple changeset with a single createTable directive with a 63-character length tableName Liquibase fails with a relation already exists error.  Not only is this error misleading, it also is unnecessary since Postgres will not truncate table names until they are MORE THAN 63 characters.  We are looking at upgrading from Liquibase 3.5.4 to 4.0.0 and we have a programmatic table name generator that currently banks on being able to create table names up to and including 63 characters in length so this is a problem for us as we have existing 63 character tables.

## Steps To Reproduce
Create a simple changeset:

```
    <changeSet id="1" author="Me">
        <createTable tableName="my_63_char________________________________________________table">
            <column name="id" type="SERIAL"><constraints primaryKey="true"/></column>
            <column name="value" type="TEXT"><constraints nullable="false" /></column>
        </createTable>
    </changeSet>
```

Execute by some means (we do it programmatically via the Java API).

The operation will fail and the nested Exception root cause is 

```
liquibase.exception.DatabaseException: ERROR: relation "my_63_char________________________________________________table" already exists [Failed SQL: (0) CREATE TABLE public.my_63_char________________________________________________table (id INTEGER GENERATED BY DEFAULT AS IDENTITY NOT NULL, value TEXT NOT NULL, CONSTRAINT MY_63_CHAR________________________________________________TABLE_PKEY PRIMARY KEY (id))]
```

This is interesting so I executed that directly in psql and what is happening is that the table creation is fine but the name of the derived primary key is automatically truncated by Postgres to 63 characters which makes it identical to the table name (the `_PKEY` suffix is shaved off) and causes the confusing error message about the relation already existing.

For whatever reason this wasn't a problem in Liquibase 3.5.4 and, if I just execute a simple create table via psql:

`create table my_63_char________________________________________________table ( id serial primary key, value text not null );`

this works fine and Postgres creates the associated PK constraint and sequence using the names `my_63_char_________________________________________________pkey` and `my_63_char_______________________________________________id_seq`.

My suspicion is that Liquibase 4.0.0 is doing it's own name creation for things like the PK constraint but that it's simply suffixing rather than first truncating to (63 - suffix_length) characters and then suffixing.

## Actual Behavior
see above

## Expected/Desired Behavior
see above

## Screenshots (if appropriate)
n/a

## Additional Context
see above

## **LIQUIBASE INTERNAL QA REQUIREMENTS**
Validate you can reproduce the bug using the steps provided prior to testing the fix.

##### Test Cases
* Verify update of a table with a 63-character name that defines a primary key on a column is successful. 
* Verify the primary key created has a name that is 63 characters long, including the suffix.
* Verify an update the 63-character name table with a primary key defined in a second changeset is successful.
* Verify all automated functional tests on Jenkins are passing
##### Note to Community: The functional Liquibase tests run on our internal build system and are not visible to the broader community.

* When manual tests and existing automated tests are passing, add a regression test to the functional test suite.
  * The test is postgres-specific and should be added as a regression test.

##### The Associated PR: [https://github.com/liquibase/liquibase/pull/1514](https://github.com/liquibase/liquibase/pull/1514)

┆Issue is synchronized with this [Jira Bug](https://datical.atlassian.net/browse/LB-827) by [Unito](https://www.unito.io/learn-more)
┆Fix Versions: Community 4.4.0,Liquibase 4.4.0

**Comment by molivasdat** [comment]
Hi @jamey-clari Thank for the excellent explanation of the issue.  We will add this to our list of issues.  If you have a PR that will make it work better that would be great.

**Comment by jamey-clari** [comment]
Hey @molivasdat - I don't have a PR and am unfamiliar with your codebase BUT if you are willing to point me in the right place to look in your code for the naming of structures associated with tables I'd be happy to give it a whack as I think I have a pretty good idea of what should be done.

**Comment by jamey-clari** [comment]
Here's some code for the `PostgresDatabase.generatePrimaryKeyName` that I *THINK* is to blame here.  It was changed with version 3.6 (we are upgrading from 3.5).  The tricky bits here are that Postgres caps the identifier length at a byte count and NOT a character count so the encoding used by the DB matters.  Additionally it's possible (though unlikely) for Postgres to be compiles with other than a 63 byte limit for identifiers.  I don't know how much you folks worry about such things and I don't know if you have a way to get the encoding from the DB or any metadata you have captured.  I hope this helps someone get a running start at this.  Thanks!

```
    @Test
    public void testGenerateDerivedNameLazy() {
        assertEquals(generateDerivedNameLazy("my_63_char________________________________________________table",  "_PKEY", 63), "MY_63_CHAR_________________________________________________PKEY");
        assertEquals(generateDerivedNameLazy( "my_63_char_______________________________________________table",  "_PKEY", 63), "MY_63_CHAR_______________________________________________T_PKEY");
        assertEquals(generateDerivedNameLazy(  "my_63_char______________________________________________table",  "_PKEY", 63), "MY_63_CHAR______________________________________________TA_PKEY");
        assertEquals(generateDerivedNameLazy(   "my_63_char_____________________________________________table",  "_PKEY", 63), "MY_63_CHAR_____________________________________________TAB_PKEY");
        assertEquals(generateDerivedNameLazy(    "my_63_char____________________________________________table",  "_PKEY", 63), "MY_63_CHAR____________________________________________TABL_PKEY");
        assertEquals(generateDerivedNameLazy(     "my_63_char___________________________________________table",  "_PKEY", 63), "MY_63_CHAR___________________________________________TABLE_PKEY");
        assertEquals(generateDerivedNameLazy(     "my_63_char___________________________________________table", "_PKEY2", 63), "MY_63_CHAR___________________________________________TABL_PKEY2");
        assertEquals(generateDerivedNameLazy(     "my_63_char___________________________________________table",  "_PKEY", 53),           "MY_63_CHAR_______________________________________PKEY");
    }

    @Test
    public void testGenerateDerivedNameStrict() {
        //System.out.println("JAMEY: " + "\u03A0\u03A1\u03A3\u03A6\u03A9");
        final Charset UTF8 = StandardCharsets.UTF_8;
        assertEquals(generateDerivedNameStrict("my_63_char________________________________________________\u03A0\u03A1\u03A3\u03A6\u03A9",  "_PKEY", 63, UTF8),                     "MY_63_CHAR_________________________________________________PKEY");
        assertEquals(generateDerivedNameStrict( "my_63_char_______________________________________________\u03A0\u03A1\u03A3\u03A6\u03A9",  "_PKEY", 63, UTF8),                      "MY_63_CHAR________________________________________________PKEY");
        assertEquals(generateDerivedNameStrict(  "my_63_char______________________________________________\u03A0\u03A1\u03A3\u03A6\u03A9",  "_PKEY", 63, UTF8),                 "MY_63_CHAR______________________________________________\u03A0_PKEY");
        assertEquals(generateDerivedNameStrict(   "my_63_char_____________________________________________\u03A0\u03A1\u03A3\u03A6\u03A9",  "_PKEY", 63, UTF8),                  "MY_63_CHAR_____________________________________________\u03A0_PKEY");
        assertEquals(generateDerivedNameStrict(    "my_63_char____________________________________________\u03A0\u03A1\u03A3\u03A6\u03A9",  "_PKEY", 63, UTF8),             "MY_63_CHAR____________________________________________\u03A0\u03A1_PKEY");
        assertEquals(generateDerivedNameStrict(     "my_63_char___________________________________________\u03A0\u03A1\u03A3\u03A6\u03A9",  "_PKEY", 63, UTF8),              "MY_63_CHAR___________________________________________\u03A0\u03A1_PKEY");
        assertEquals(generateDerivedNameStrict(      "my_63_char__________________________________________\u03A0\u03A1\u03A3\u03A6\u03A9",  "_PKEY", 63, UTF8),         "MY_63_CHAR__________________________________________\u03A0\u03A1\u03A3_PKEY");
        assertEquals(generateDerivedNameStrict(       "my_63_char_________________________________________\u03A0\u03A1\u03A3\u03A6\u03A9",  "_PKEY", 63, UTF8),          "MY_63_CHAR_________________________________________\u03A0\u03A1\u03A3_PKEY");
        assertEquals(generateDerivedNameStrict(        "my_63_char________________________________________\u03A0\u03A1\u03A3\u03A6\u03A9",  "_PKEY", 63, UTF8),     "MY_63_CHAR________________________________________\u03A0\u03A1\u03A3\u03A6_PKEY");
        assertEquals(generateDerivedNameStrict(         "my_63_char_______________________________________\u03A0\u03A1\u03A3\u03A6\u03A9",  "_PKEY", 63, UTF8),      "MY_63_CHAR_______________________________________\u03A0\u03A1\u03A3\u03A6_PKEY");
        assertEquals(generateDerivedNameStrict(          "my_63_char______________________________________\u03A0\u03A1\u03A3\u03A6\u03A9",  "_PKEY", 63, UTF8), "MY_63_CHAR______________________________________\u03A0\u03A1\u03A3\u03A6\u03A9_PKEY");
        assertEquals(generateDerivedNameStrict(           "my_63_char_____________________________________\u03A0\u03A1\u03A3\u03A6\u03A9",  "_PKEY", 63, UTF8),  "MY_63_CHAR_____________________________________\u03A0\u03A1\u03A3\u03A6\u03A9_PKEY");
        assertEquals(generateDerivedNameStrict(          "my_63_char______________________________________\u03A0\u03A1\u03A3\u03A6\u03A9", "_PKEY2", 63, UTF8),      "MY_63_CHAR______________________________________\u03A0\u03A1\u03A3\u03A6_PKEY2");
        assertEquals(generateDerivedNameStrict(          "my_63_char______________________________________\u03A0\u03A1\u03A3\u03A6\u03A9",  "_PKEY", 53, UTF8),  "MY_63_CHAR_______________________________________PKEY");
    }

    public String generatePrimaryKeyName(final String tableName) {
        // PG identifiers have a max length in bytes, not characters
        // Can we obtain this from the DB somehow?
        final int pgIdentifierMaxBytes = 63;

        // just assume single-byte characters for table names
        return generateDerivedNameLazy(tableName, "_PKEY", pgIdentifierMaxBytes);

        // allow for multibyte character table names
        // Can we obtain the DB charset from the DB somehow?  Else I suppose we just use the JVM platform default?
        //return generateDerivedNameStrict(tableName, "_PKEY", pgIdentifierMaxBytes, Charset.defaultCharset());
    }

    private String generateDerivedNameLazy(final String tableName,
                                           final String suffix,
                                           final int pgIdentifierMaxBytes) {
        // assume single-byte characters
        final int maxLen = pgIdentifierMaxBytes - suffix.length();
        final String truncTableName = (tableName.length() > maxLen) ? tableName.substring(0, maxLen) : tableName;
        return truncTableName.toUpperCase(Locale.US) + suffix;
    }

    private String generateDerivedNameStrict(final String tableName,
                                             final String suffix,
                                             final int pgIdentifierMaxBytes,
                                             final Charset dbCharset) {
        final byte[] suffixBytes = suffix.getBytes(dbCharset);
        final int maxLen = pgIdentifierMaxBytes - suffixBytes.length;
        String truncTableName = tableName.toUpperCase(Locale.US);
        while (truncTableName.getBytes(dbCharset).length > maxLen) {
            truncTableName = truncTableName.substring(0, truncTableName.length()-1);
        }
        return truncTableName + suffix;
    }
```

**Comment by jamey-clari** [comment]
Some dabbling in PG 10 shows that the behavior above should be doing a good job of emulating the PG truncation behavior.  Note that Postgres dowcases all such identifiers too so the `toUpperCase` performed by generateDerivedName is seemingly pointless.

```
my_db=> create table MY_63_CHAR________________________________________________ΠΡΣΦΩ ( id serial primary key, value text not null );
NOTICE:  identifier "my_63_char________________________________________________ΠΡΣΦΩ" will be truncated to "my_63_char________________________________________________ΠΡ"
CREATE TABLE
my_db=> \d "my_63_char________________________________________________ΠΡ";
                          Table "public.my_63_char________________________________________________ΠΡ"
 Column |  Type   | Collation | Nullable |                                       Default                                        
--------+---------+-----------+----------+--------------------------------------------------------------------------------------
 id     | integer |           | not null | nextval('my_63_char_______________________________________________id_seq'::regclass)
 value  | text    |           | not null | 
Indexes:
    "my_63_char_________________________________________________pkey" PRIMARY KEY, btree (id)
my_db=> drop table  my_63_char________________________________________________ΠΡ;
DROP TABLE
```

**Comment by molivasdat** [comment]
Thanks for the additional research on PostgreSQL 10.

**Comment by alexey-semenov** [comment]
Could I work on it?
Also I have a question. Postgres DB PK names is built using `_PKEY` suffix but it's `PK_` prefix for other DBs. Is that expected? Or just left for some backward compatibility? Could you please clarify. If there're no any relevant reasons to keep it as suffix for Postgres then I'd propose to go with the same approach as it's done for other DBs  and use `PK_` prefix to avoid truncating key names. Thanks.

**Comment by jamey-clari** [comment]
It'd be great to have it resolved by some means so, speaking as a grateful user, thanks for offering!  I myself am new to the project other than working on an application using v3.5 of the Java library.  It appears that the override on the PK naming routine (`PostgresDatabase.generatePrimaryKeyName`) came in with v3.6.  I was assuming that it was determined that emulating the Postgres convention was desirable for some reason.  Unfortunately the Postgres long-name truncation logic wasn't replicated.  An alternative would be to NOT explicitly name PKs in Liquibase using Postgres, if that were an option, and let Postgres apply its convention implicitly.  I suspect though that Liquibase needs to be aware of the PK naming for some purpose though and thus the presence of that function in the APIs and explicit naming by Liquibase in table create statements.  Library authors will presumably know what to do here where as I can merely speculate on the history and intentions.

**Comment by alexey-semenov** [comment]
@jamey-clari Yeah. I've just double checked it. In accordance with Postgres naming conventions PKs should be suffixed with `_pkey`. So, if no objections, will take a deeper look and try to fix it.

**Comment by jamey-clari** [comment]
Awesome, thanks @alexey-semenov !

**Comment by alexey-semenov** [comment]
Guys, so sorry if I'm not following all your git-flow conventions. 
Let me know if I should correct any descriptions or some formatting.
Here is MR I think should fix the issue with truncating PKs https://github.com/alexey-semenov/liquibase/pull/new/issue-1483
Please review.

**Comment by jamey-clari** [comment]
May I request that this fix be applied to versions 3.6 and forward rather than just 4.0 and forward?  The change which caused the naming problem for postgres (the override of PostgresDatabase.generatePrimaryKeyName) was done in 3.6.  Thank you for considering it.

**Comment by alexey-semenov** [comment]
@molivasdat @jamey-clari Guys, that should be fine now. Got rid of hardcoded encoding and added non-ASCII names test. Please review.

**Comment by molivasdat** [comment]
Hi @jamey-clari What is the hesitation with moving to the 4.x release for your team?

**Comment by jamey-clari** [comment]
@molivasdat - I don't have an explicit one at the moment.  Most libraries have breaking API changes in major point releases so I don't know how many of those I might run into and need to overcome as I've not done extensive testing of 4.x yet.  If the hurdles were significant I might elect an intermediate point of upgrade for the time being and depending on other pressures on me.  I did stub my toe on the naming issue here and, admittedly, I think that was all I stubbed my toe on in our unit test suite after making some code adjustments for the 4.x APIs.  Given that, PERHAPS I'm actually pretty close to being able to seriously consider 4.x.  I just haven't spent a lot of time with it yet.  Also, if you are still supporting any of the latter 3.x point releases then I would argue that this is a regression that should be addressed there too.  My two cents, thanks for listening:)

**Comment by molivasdat** [comment]
Hi @jamey-clari Thanks for the clarification.  There is not a plan to fix this in the 3.x branch.  It will be merged into the 4.x branch moving forward.  Hoping that this does not cause too much of an issue.  We are glad that you are close to moving to 4.x though.

**Comment by jamey-clari** [comment]
Thanks @molivasdat - You can't blame me for trying;)

**Comment by alexey-semenov** [comment]
Guys, sorry for being absent for a long time. Was up to my years in work. Pushed updates, please check.

**Comment by molivasdat** [comment]
@jamey-clari No blame here. It's all good.

**Comment by jamey-clari** [comment]
@molivasdat - Any idea when this might find its way into the release pipeline?  We'd like to upgrade to 4.x and this is currently blocking us.  Thanks!

**Comment by molivasdat** [comment]
Thanks for letting me know @jamey-clari .  It won't make the current release which is currently in UAT but I'll make sure to move it along for the next list of issues to fix.

**Comment by mariochampion** [comment]
hey @jamey-clari -- with the pull request associated with this (thanks!), we will initially slate this for the release **after** 4.3.2. There isnt a firm timeline yet, but it is on the order or a month-ish, not multiple months.

**Comment by jamey-clari** [comment]
Wonderful, thanks @mariochampion !

**Comment by jamey-clari** [comment]
> hey @jamey-clari -- with the pull request associated with this (thanks!), we will initially slate this for the release **after** 4.3.2. There isnt a firm timeline yet, but it is on the order or a month-ish, not multiple months.

Hey folks, I'm still not seeing this in the 4.3.5 release of a few days ago.  Any updated ETA?  I'd love to be able to integrate 4.3.x into our application.  Thanks!

**Comment by molivasdat** [comment]
Hi @jamey-clari It's merged into the 4.4.0 release branch, internally.  It's nearing completion, taking a little more time than we anticipated.  I'll let @mariochampion comment on release.

**Comment by jamey-clari** [comment]
Thanks @molivasdat !
